### ISSUE TITLE
[BUG] binops downcast inputs incorrectly before operation

### ISSUE THREAD

**Issue by jlowe** [rapidsai/cudf#2574]
**Describe the bug**
When performing a division of types and requesting a downcasted type as the output, the downcast occurs on the operations _before_ the division occurs resulting in a loss of precision and unexpected output.

**Steps/Code to reproduce bug**
Apply this patch to binary-operation-integration-test.cpp and run the `BINARY_TEST` gtest:
```patch
diff --git a/cpp/tests/binaryop/integration/binary-operation-integration-test.cpp b/cpp/tests/binaryop/integration/binary-operation-integration-test.cpp
index 020ea933f..b83e0ceea 100644
--- a/cpp/tests/binaryop/integration/binary-operation-integration-test.cpp
+++ b/cpp/tests/binaryop/integration/binary-operation-integration-test.cpp
@@ -141,6 +141,24 @@ TEST_F(BinaryOperationIntegrationTest, Div_Vector_Vector_SI64) {
 }
 
 
+TEST_F(BinaryOperationIntegrationTest, Div_Vector_Vector_Downcast) {
+    auto lhs = cudf::test::column_wrapper<int64_t>(1,
+        [](gdf_size_type row) {return 0x12345678;},
+        [](gdf_size_type row) {return true;});
+    auto rhs = cudf::test::column_wrapper<int64_t>(1,
+        [](gdf_size_type row) {return 0x10;},
+        [](gdf_size_type row) {return true;});
+    auto out = cudf::test::column_wrapper<int8_t>(lhs.get()->size, true);
+
+    CUDF_EXPECT_NO_THROW(cudf::binary_operation(out.get(), lhs.get(), rhs.get(), GDF_DIV));
+
+    auto out_h = out.to_host();
+    auto out_data = std::get<0>(out_h);
+    ASSERT_TRUE(out_data.size() == 1);
+    ASSERT_EQ(out_data[0], 0x67);
+}
+
+
```

**Expected behavior**
The downcast would occur after the division operation to preserve precision.

**Comment by devavret** [comment]
For your use case, try `binary_operator::TRUE_DIV`. It casts each operand to `double`, divides and casts back to output type. `binary_operator::DIV` probably exists for old python compatibility but we can either do away with it entirely or change it to C style div i.e. simply `lhs / rhs`

**Comment by jlowe** [comment]
TRUE_DIV is what I ended up using.  I went ahead and filed this issue because I found the behavior unexpected.  I assumed incorrectly that the specified output type would be applied intelligently, i.e.: preserving as much precision as possible during the operation.

If binops always casts the output type on both inputs before performing the operation regardless of the operation type then I suppose this is "working as intended."  Users will have to be careful when specifying a less precise type for outputs, hoping to get a "free" downcast after the operation.
